pkg/gardenlet/controller/shoot/shoot/reconciler_force_delete.go

```diff
@@ -0,0 +1,193 @@
+			o.CleanShootTaskError(ctx, errorID)
+			return nil
+		},
+		nil,
+		errors.ToExecute("Create botanist", func() error {
+			return retryutils.UntilTimeout(ctx, 10*time.Second, 10*time.Minute, func(context.Context) (done bool, err error) {
+				botanist, err = botanistpkg.New(ctx, o)
+				if err != nil {
+					return retryutils.MinorError(err)
+				}
+				return retryutils.Ok()
+			})
+		}),
+		errors.ToExecute("Check required extensions exist", func() error {
+			return botanist.WaitUntilRequiredExtensionsReady(ctx)
+		}),
+		// We first check whether the namespace in the Seed cluster does exist - if it does not, then we assume that
+		// all resources have already been deleted. We can delete the Shoot resource as a consequence.
+		errors.ToExecute("Retrieve the Shoot namespace in the Seed cluster", func() error {
+			return checkIfSeedNamespaceExists(ctx, o, botanist)
+		}),
+	)
+
+	if err != nil {
+		return v1beta1helper.NewWrappedLastErrors(v1beta1helper.FormatLastErrDescription(err), err)
+	}
+
+	var (
+		defaultInterval = 5 * time.Second
+		defaultTimeout  = 30 * time.Second
+
+		cleaner = NewCleaner(botanist.SeedClientSet.Client(), r.GardenClient, botanist.Shoot.SeedNamespace, botanist.Shoot.GetInfo().Namespace, botanist.Shoot.BackupEntryName, logf.FromContext(ctx))
+
+		g = flow.NewGraph("Shoot cluster force deletion")
+
+		deleteExtensionObjects = g.Add(flow.Task{
+			Name: "Deleting extension resources",
+			Fn:   flow.TaskFn(cleaner.DeleteExtensionObjects).RetryUntilTimeout(defaultInterval, defaultTimeout),
+		})
+		waitUntilExtensionObjectsDeleted = g.Add(flow.Task{
+			Name:         "Waiting until extension resources have been deleted",
+			Fn:           cleaner.WaitUntilExtensionObjectsDeleted,
+			Dependencies: flow.NewTaskIDs(deleteExtensionObjects),
+		})
+		deleteMCMResources = g.Add(flow.Task{
+			Name: "Deleting MCM resources",
+			Fn:   flow.TaskFn(cleaner.DeleteMCMResources).RetryUntilTimeout(defaultInterval, defaultTimeout),
+		})
+		waitUntilMCMResourcesDeleted = g.Add(flow.Task{
+			Name:         "Waiting until MCM resources have been deleted",
+			Fn:           cleaner.WaitUntilMCMResourcesDeleted,
+			Dependencies: flow.NewTaskIDs(deleteMCMResources),
+		})
+		deleteCluster = g.Add(flow.Task{
+			Name:         "Deleting Cluster resource",
+			Fn:           flow.TaskFn(cleaner.DeleteCluster).RetryUntilTimeout(defaultInterval, defaultTimeout),
+			Dependencies: flow.NewTaskIDs(waitUntilExtensionObjectsDeleted),
+		})
+		waitUntilClusterDeleted = g.Add(flow.Task{
+			Name:         "Waiting until Cluster resource has been deleted",
+			Fn:           cleaner.WaitUntilClusterDeleted,
+			Dependencies: flow.NewTaskIDs(deleteCluster),
+		})
+		setKeepObjectsForManagedResources = g.Add(flow.Task{
+			Name: "Configuring managed resources to not keep their objects when deleted",
+			Fn: flow.TaskFn(flow.TaskFn(func(ctx context.Context) error {
+				return botanist.SetKeepObjectsForManagedResources(ctx, false)
```

Comment:
This will only set the `keepObjects` field for `ManagedResources` deployed by gardener (MRs that have the `gardener` origin label). However, in subsequent steps you will try to delete all `ManagedResources`.

In the migration flow we let all extensions migrate their own `ManagedResources`. There's one step at the end of the flow which waits for ALL `ManagedResources` to be deleted: https://github.com/gardener/gardener/blob/b9651c20c89d4b19a7615873d56f2aa71381e624/pkg/gardenlet/controller/shoot/shoot/reconciler_migrate.go#L259-L263